### PR TITLE
Dataset string memcap 3910/v6

### PR DESCRIPTION
**Pull request by inashivb**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3910

Previous PR: https://github.com/OISF/suricata/pull/11105

Changes since v5:
- doc suggestion taken into account
- reorganize commits better so it requires less changes while still making sense logically
- rebased
- minor cleanups whitespaces et al.

### CODE REVIEW

**Comment by ct0br0** [comment]
qa tokens ran out. ignore the 'pipeline canceled' incoming

File: src/util-thash.c [hunk 1]
```diff
@@ -371,6 +378,8 @@ void THashShutdown(THashTableContext *ctx)
     }
     (void) SC_ATOMIC_SUB(ctx->memuse, ctx->config.hash_size * sizeof(THashHashRow));
     THashDataQueueDestroy(&ctx->spare_q);
+    SCLogDebug("memuse: %ld", (uint64_t)SC_ATOMIC_GET(ctx->memuse));
```
Comment (victorjulien) [inline src/util-thash.c:381]: our pattern is to use`"%"PRIu64` for `uint64_t`

File: src/util-thash.c [hunk 2]
```diff
@@ -516,7 +581,10 @@ static THashData *THashDataGetNew(THashTableContext *ctx, void *data)
 
     // setup the data
     BUG_ON(ctx->config.DataSet(h->data, data) != 0);
-
+    if (ctx->config.DataSize) {
+        uint32_t len = ctx->config.DataSize(data);
+        (void)SC_ATOMIC_ADD(ctx->memuse, (uint64_t)len);
```
Comment (victorjulien) [inline src/util-thash.c:586]: where do we actually check if the data we intend to store fits the memcap?

Comment (inashivb) [reply src/util-thash.c:586]: Every time a dataset is added, a call is made to `THashGetFromHash` which returns either the found `THashData` or creates new data using `THashDataGetNew`  <-- this checks for `memcap`.

After a dataset has been added, there's a call made to `THashConsolidateMemcap` <- this fn tends to update the `memcap` as per the current `memuse`. This fn is the reason, I chose to increment `memuse` as the rest should be taken care of as it is for datasets of the other types as well.
wdyt?

Comment (victorjulien) [reply src/util-thash.c:586]: But we shouldn't add data if it violates the memcap, and it seems we're not checking.

Comment (inashivb) [reply src/util-thash.c:586]: indeed. The memuse update is happening post addition. Not sure why did I not see it. 😓

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20779

**Review by jufajardini — COMMENTED** [review]
Doc change looks good to me :P

**Pull request closed without merging**